src/github.com/travis-ci/worker/image/api_selector.go

```diff
@@ -0,0 +1,181 @@
+package image
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/cenkalti/backoff"
+)
+
+const (
+	imageAPIRequestContentType = "application/x-www-form-urlencoded; boundary=NL"
+)
+
+type APISelector struct {
+	baseURL *url.URL
+
+	maxInterval    time.Duration
+	maxElapsedTime time.Duration
+}
+
+func NewAPISelector(u *url.URL) *APISelector {
+	return &APISelector{
+		baseURL: u,
+
+		maxInterval:    10 * time.Second,
+		maxElapsedTime: time.Minute,
+	}
+}
+
+func (as *APISelector) Select(params *Params) (string, error) {
+	imageName, err := as.queryWithTags(params.Infra, as.buildCandidateTags(params))
+	if err != nil {
+		return "default", err
+	}
+
+	if imageName != "" {
+		return imageName, nil
+	}
+
+	return "default", nil
+}
+
+func (as *APISelector) queryWithTags(infra string, tags [][]string) (string, error) {
+	bodyLines := []string{}
+
+	for _, ts := range tags {
+		qs := url.Values{}
+		qs.Set("infra", infra)
+		qs.Set("fields[images]", "name")
+		qs.Set("limit", "1")
+		if len(tags) > 0 {
+			qs.Set("tags", strings.Join(ts, ","))
+		}
+
+		bodyLines = append(bodyLines, qs.Encode())
+	}
+
+	u, _ := url.Parse(as.baseURL.String())
+	imageResp, err := as.makeImageRequest(u.String(), bodyLines)
+	if err != nil {
+		return "", err
+	}
+
+	if len(imageResp.Data) > 0 {
```

Comment:
This is more of a style thing, but I remember reading something somewhere about keeping the "happy path" at least indentation, so I'm wondering if it would be a good idea to invert this `if`, so that the `return "", nil` error condition is inside the braces and the `return imageResp.Data[0].Name, nil` is at the bottom of the method?